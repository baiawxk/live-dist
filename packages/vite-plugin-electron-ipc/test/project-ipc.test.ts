import { promises as fs } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import plugin from '../src/index'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

// 使用项目中实际的 IPC 文件进行测试
const projectIpcDir = resolve(__dirname, '../../main/src/ipc')
const projectMainRegistryFile = resolve(__dirname, '../../main/src/generated/ipc-registry.ts')
const projectPreloadClientFile = resolve(__dirname, '../../preload/src/generated/ipc-client.ts')
const projectTypeDefFile = resolve(__dirname, '../../preload/src/generated/ipc-types.d.ts')

describe('iPC Plugin - Real Project Files', () => {
  it('should scan real project IPC functions correctly', async () => {
    const ipcPlugin = plugin({
      scanDir: projectIpcDir,
      mainRegistryFile: projectMainRegistryFile,
      preloadClientFile: projectPreloadClientFile,
      typeDefinitionFile: projectTypeDefFile,
    })

    // 模拟配置解析
    await ipcPlugin.configResolved({ mode: 'development' })
    
    // 模拟构建开始
    await ipcPlugin.buildStart()

    // 检查生成的文件是否存在
    const mainRegistryExists = await fs.access(projectMainRegistryFile).then(() => true).catch(() => false)
    const preloadClientExists = await fs.access(projectPreloadClientFile).then(() => true).catch(() => false)
    const typeDefExists = await fs.access(projectTypeDefFile).then(() => true).catch(() => false)

    expect(mainRegistryExists).toBe(true)
    expect(preloadClientExists).toBe(true)
    expect(typeDefExists).toBe(true)
  })

  it('should generate main registration code for real project', async () => {
    const ipcPlugin = plugin({
      scanDir: projectIpcDir,
      mainRegistryFile: projectMainRegistryFile,
      preloadClientFile: projectPreloadClientFile,
      typeDefinitionFile: projectTypeDefFile,
    })

    // 模拟配置解析
    await ipcPlugin.configResolved({ mode: 'development' })
    
    // 模拟构建开始
    await ipcPlugin.buildStart()

    // 检查生成的主进程注册代码
    const mainCode = await fs.readFile(projectMainRegistryFile, 'utf-8')
    expect(mainCode).toContain('import { openInBrowser, selectAppDirectory } from')
    expect(mainCode).toContain('import { getAllDists, addDist, updateDist, removeDist, getDist, updateDistStatus, selectDistDirectory } from')
    expect(mainCode).toContain('import { startServer, stopServer, getServerStatus } from')
    expect(mainCode).toContain('export function registerIPCFunctions()')
  })

  it('should generate preload client code for real project', async () => {
    const ipcPlugin = plugin({
      scanDir: projectIpcDir,
      mainRegistryFile: projectMainRegistryFile,
      preloadClientFile: projectPreloadClientFile,
      typeDefinitionFile: projectTypeDefFile,
    })

    // 模拟配置解析
    await ipcPlugin.configResolved({ mode: 'development' })
    
    // 模拟构建开始
    await ipcPlugin.buildStart()

    // 检查生成的preload客户端代码
    const preloadCode = await fs.readFile(projectPreloadClientFile, 'utf-8')
    expect(preloadCode).toContain('import { ipcRenderer } from \'electron\'')
    expect(preloadCode).toContain('export const app = {')
    expect(preloadCode).toContain('export const distMgr = {')
    expect(preloadCode).toContain('export const server = {')
  })

  it('should generate type definitions for real project', async () => {
    const ipcPlugin = plugin({
      scanDir: projectIpcDir,
      mainRegistryFile: projectMainRegistryFile,
      preloadClientFile: projectPreloadClientFile,
      typeDefinitionFile: projectTypeDefFile,
    })

    // 模拟配置解析
    await ipcPlugin.configResolved({ mode: 'development' })
    
    // 模拟构建开始
    await ipcPlugin.buildStart()

    // 检查生成的类型定义文件
    const typeDefs = await fs.readFile(projectTypeDefFile, 'utf-8')
    expect(typeDefs).toContain('export const app: {')
    expect(typeDefs).toContain('export const distMgr: {')
    expect(typeDefs).toContain('export const server: {')
    
    // 验证类型定义包含项目中的实际函数
    expect(typeDefs).toContain('openInBrowser: typeof import(')
    expect(typeDefs).toContain('getAllDists: typeof import(')
    expect(typeDefs).toContain('startServer: typeof import(')
  })
})
