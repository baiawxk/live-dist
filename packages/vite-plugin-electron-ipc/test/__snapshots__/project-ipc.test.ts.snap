// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`iPC Plugin - Real Project Files > should generate main registration code for real project 1`] = `
"
import { openInBrowser } from 'E:/workspaces/live-dist/packages/main/src/ipc/app.ts';
import { selectAppDirectory } from 'E:/workspaces/live-dist/packages/main/src/ipc/app.ts';
import { getAllDists } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { addDist } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { updateDist } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { removeDist } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { getDist } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { updateDistStatus } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { selectDistDirectory } from 'E:/workspaces/live-dist/packages/main/src/ipc/distMgr.ts';
import { startServer } from 'E:/workspaces/live-dist/packages/main/src/ipc/server.ts';
import { stopServer } from 'E:/workspaces/live-dist/packages/main/src/ipc/server.ts';
import { getServerStatus } from 'E:/workspaces/live-dist/packages/main/src/ipc/server.ts';
import { ipcMain } from 'electron';

export function registerIPCFunctions() {
  ipcMain.handle('app:openInBrowser', (_, ...args) => openInBrowser(...args));
ipcMain.handle('app:selectAppDirectory', (_, ...args) => selectAppDirectory(...args));
ipcMain.handle('distMgr:getAllDists', (_, ...args) => getAllDists(...args));
ipcMain.handle('distMgr:addDist', (_, ...args) => addDist(...args));
ipcMain.handle('distMgr:updateDist', (_, ...args) => updateDist(...args));
ipcMain.handle('distMgr:removeDist', (_, ...args) => removeDist(...args));
ipcMain.handle('distMgr:getDist', (_, ...args) => getDist(...args));
ipcMain.handle('distMgr:updateDistStatus', (_, ...args) => updateDistStatus(...args));
ipcMain.handle('distMgr:selectDistDirectory', (_, ...args) => selectDistDirectory(...args));
ipcMain.handle('server:startServer', (_, ...args) => startServer(...args));
ipcMain.handle('server:stopServer', (_, ...args) => stopServer(...args));
ipcMain.handle('server:getServerStatus', (_, ...args) => getServerStatus(...args));
}
"
`;

exports[`iPC Plugin - Real Project Files > should generate preload client code for real project 1`] = `
"
import { ipcRenderer } from 'electron'

// 按文件名分组的API对象
export const app = {
  openInBrowser: (...args) => ipcRenderer.invoke('app:openInBrowser', ...args),
  selectAppDirectory: (...args) => ipcRenderer.invoke('app:selectAppDirectory', ...args)
};

export const distMgr = {
  getAllDists: (...args) => ipcRenderer.invoke('distMgr:getAllDists', ...args),
  addDist: (...args) => ipcRenderer.invoke('distMgr:addDist', ...args),
  updateDist: (...args) => ipcRenderer.invoke('distMgr:updateDist', ...args),
  removeDist: (...args) => ipcRenderer.invoke('distMgr:removeDist', ...args),
  getDist: (...args) => ipcRenderer.invoke('distMgr:getDist', ...args),
  updateDistStatus: (...args) => ipcRenderer.invoke('distMgr:updateDistStatus', ...args),
  selectDistDirectory: (...args) => ipcRenderer.invoke('distMgr:selectDistDirectory', ...args)
};

export const server = {
  startServer: (...args) => ipcRenderer.invoke('server:startServer', ...args),
  stopServer: (...args) => ipcRenderer.invoke('server:stopServer', ...args),
  getServerStatus: (...args) => ipcRenderer.invoke('server:getServerStatus', ...args)
};

"
`;
